drivers: led: Microchip XEC LED driver using BBLED controller

jvasanth1 opened the pull request:
Implement a LED driver for Microchip XEC using the breathing, blinking LED controller. The driver supports LED on, off, and
blink API's.

jvasanth1 committed:
samples: drivers: led: Add sample for Microchip XEC LED driver

Add a sample program demonstrating the LED driver for Microchip's
XEC microcontrollers. The sample can be built for MEC152x and
MEC172x.

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

simonguinot commented:
Hi @jvasanth1. Thanks for this driver. I'll review it this week-end.

albertofloyd commented on drivers/led/Kconfig.xec:
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2021 Nordic Semiconductor ASA
```
Microchip?

jvasanth1 replied:
fixed

albertofloyd commented:
LGTM 
@jvasanth1 looks like maybe there are some .yaml files needed to be update for CI @nashif any pointer here?

simonguinot commented on drivers/led/led_mchp_xec.c:
```diff
@@ -0,0 +1,292 @@
+static uint32_t calc_blink_32k_prescaler(uint32_t delay_on, uint32_t delay_off)
+{
+	uint32_t temp = ((delay_on + delay_off) * XEC_BBLED_BLINK_CLK_SRC_HZ) / (256U * 1000U);
+	uint32_t prescaler = 0u;
```
You can drop 'u' here.

jvasanth1 replied:
updated

simonguinot commented on samples/drivers/led_xec/README.rst:
```diff
@@ -0,0 +1,75 @@
+********
+
+This sample allows to test the Microchip led-xec driver which uses the
+breathing-blinking LED (BBLED) controllers. The SoC is design is fixed
```
There are too many "is" above.

jvasanth1 replied:
updated

simonguinot reviewed (commented):
Hi @jvasanth1,

Thanks for this driver. It is looking good to me.

Please see below for few questions and comments I have.

jvasanth1 commented on dts/bindings/led/microchip,xec-bbled.yaml:
```diff
@@ -0,0 +1,31 @@
+
+description: Microchip XEC BBLED
+
+include: [led-controller.yaml, base.yaml, pinctrl-device.yaml]
```
okay, we will try removing it

jvasanth1 replied:
updated

jvasanth1 commented on drivers/led/led_mchp_xec.c:
```diff
@@ -0,0 +1,292 @@
+	.regs = (struct xec_bbled_regs * const)DT_INST_REG_ADDR(i),	\
+	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(i),			\
+	.girq_id = (uint8_t)(DT_INST_PROP_BY_IDX(i, girqs, 0)),		\
+	.girq_pos = (uint8_t)(DT_INST_PROP_BY_IDX(i, girqs, 1)),	\
```
we will remove them.

jvasanth1 replied:
updated

jvasanth1 commented on drivers/led/Kconfig.xec:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config LED_XEC
+	bool "Microchip XEC BB-LED driver"
```
XEC is what Intel recommended us to name all Microchip based Embedded Controllers

jvasanth1 commented on drivers/led/led_mchp_xec.c:
```diff
@@ -0,0 +1,292 @@
+	regs->config |= BIT(XEC_BBLED_CFG_RST_PWM_POS);
+	regs->config = XEC_BBLED_CFG_MODE_OFF;
+
+	int ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
```
pinctrl is mandatory now

jvasanth1 replied:
updated `ret`

jvasanth1 commented on drivers/led/led_mchp_xec.c:
```diff
@@ -0,0 +1,292 @@
+static const struct led_driver_api xec_bbled_api = {
+	.on		= xec_bbled_on,
+	.off		= xec_bbled_off,
+	.blink		= xec_bbled_blink,
```
Currently we are not supporting it because we don't see any corresponding Zephyr API for breathing feature.

simonguinot replied:
Well. It may be the opportunity to add one :) Please could you help to design a such API ? Is there some documents describing how the hardware breathing feature is configured ?

jvasanth1 replied:
We want this PR to get approved and merged first, since this basic bbled driver is required by our customers.
Adding breathing LED feature will involve discussions and will be a separate PR.

simonguinot replied:
OK I understand. The PR is good. So I am confident it will be merged soon. BTW don't forget to answer my other questions if you want my approval.

Thanks to your customers, you are sponsored to work on this driver. So it seems to me it is the perfect timing to contribute a new feature instead of just dropping by a driver. After you will move to something else and the opportunity will be over. 

Note that the LED breathing API will be nothing more than a wrapping function with a `struct led_breathing` passed as argument. Probably this structure will include information such as the ramp up, ramp down, down and up delays, the minimum and maximum brightness, and eventually the initial state to start the effect. That's it. It will be a 50 lines or so patch. Everything else will be the specific implementation in your driver for your device. This will a have a real value for your customers.

jvasanth1 replied:
could you help to approve this PR? I think I have responded to all your comments.

simonguinot replied:
Sure, we are almost there. So, how do you feel about adding a LED breathing API and a nice feature for your customers ?

jvasanth1 replied:
We will consider adding the breathing API in a separate PR.

jvasanth1 commented on dts/arm/microchip/mec1501hsz.dtsi:
```diff
@@ -489,6 +489,27 @@
 			#address-cells = <1>;
 			#size-cells = <0>;
 		};
+		bbled0: bbled@4000b800 {
```
okay will create separate commit for MEC15xx support

jvasanth1 replied:
updated

jvasanth1 commented on drivers/led/led_mchp_xec.c:
```diff
@@ -0,0 +1,292 @@
+	const struct xec_bbled_config * const config = dev->config;
+	struct xec_bbled_regs * const regs = config->regs;
+
+	xec_bbled_slp_en_clr(dev);
```
we will consider it later.

simonguinot replied:
Just been curious and looking more at this, it looks like that the chip is able to operate at low power mode when the 32KHz clock is used (which seems to be always the case here) and when the SLEEP_ENABLE bits in the PCR register are set. Are the LEDs behaving normally when operating in low power mode ?

jvasanth1 replied:
Yes, the LEDs behaves normally in low power mode.

simonguinot replied:
So shouldn't we put the chip in low power mode instead to save power since LEDs behave normally ?

jvasanth1 replied:
Currently the implementation keeps the bbled ON in low power mode since it is using the 32Khz clock. This is also as per the use case of our customer. 
We are looking into adding option through device tree to turn off the bbled hardware when entering low power; this will be a separate PR.

jvasanth1 commented on drivers/led/Kconfig.xec:
```diff
@@ -0,0 +1,10 @@
+	depends on DT_HAS_MICROCHIP_XEC_BBLED_ENABLED
+	help
+	  Enable driver for the Microchip XEC Breathing-Blinking
+	  LED controller
```
updated

jvasanth1 commented on drivers/led/led_mchp_xec.c:
```diff
@@ -0,0 +1,292 @@
+			    uint32_t delay_on, uint32_t delay_off)
+{
+	const struct xec_bbled_config * const config = dev->config;
+	struct xec_bbled_regs * const regs = config->regs;
```
skipping this

simonguinot replied:
OK

jvasanth1 committed:
drivers: led: Microchip XEC LED driver using BBLED controller

Implement a LED driver for Microchip XEC using the breathing,
blinking LED controller. The driver supports LED on, off, and
blink API's. The BBLED block uses the 32768 Hz clock domain
allowing the module to operate in light and deep sleep states.
Blink frequency is 32768 divided by 256 * (prescale + 1) where
prescale is a 12-bit value. Duty cycle is specified by an 8-bit
value where 0 = full off, 127 is 50%, and 255 is full on.

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 committed:
dts: arm: mec1501: add bbled support for MEC1501

Add bbled entries in MEC1501hsz.dts for MEC1501 bbled support

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

simonguinot commented on samples/drivers/led_xec/boards/mec15xxevb_assy6853.overlay:
```diff
@@ -0,0 +1,29 @@
+ */
+
+/ {
+};
```
What is the purpose for the parenthesis above ?

jvasanth1 replied:
not required, updated

simonguinot commented on samples/drivers/led_xec/boards/mec172xevb_assy6906.overlay:
```diff
@@ -0,0 +1,36 @@
+ */
+
+/ {
+};
```
Same here

jvasanth1 replied:
not required, updated

Mani-Sadhasivam commented on drivers/led/led_mchp_xec.c:
```diff
@@ -0,0 +1,289 @@
+/* return duty cycle scaled to [0, 255]
+ * caller must insure delay_on and delay_off are in hardware range.
+ */
+static inline uint32_t calc_blink_duty_cycle(uint32_t delay_on, uint32_t delay_off)
```
`inline` keyword in C files is not needed nowadays as the compiler will make the function inline if it sees fit.

jvasanth1 replied:
fixed

Mani-Sadhasivam commented on drivers/led/led_mchp_xec.c:
```diff
@@ -0,0 +1,289 @@
+	regs->config = XEC_BBLED_CFG_MODE_OFF;
+
+	ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+
```
No need of a newline here.

jvasanth1 replied:
fixed

Mani-Sadhasivam reviewed (changes requested):
Just minor nitpicks. Rest LGTM!

Mani-Sadhasivam commented:
@carlescufi Please merge!
